clippy_lints/src/absolute_symbol_paths.rs

```diff
@@ -0,0 +1,100 @@
+    /// ### Example
+    /// ```rust
+    /// let x = std::f64::consts::PI;
+    /// ```
+    /// Use any of the below instead, or anything else:
+    /// ```rust
+    /// # use std::f64;
+    /// # use std::f64::consts;
+    /// # use std::f64::consts::PI;
+    /// let x = f64::consts::PI;
+    /// let x = consts::PI;
+    /// let x = PI;
+    /// use std::f64::consts as f64_consts;
+    /// let x = f64_consts::PI;
+    /// ```
+    #[clippy::version = "1.72.0"]
+    pub ABSOLUTE_SYMBOL_PATHS,
+    restriction,
+    "checks for usage of a symbol without a `use` statement"
+}
+impl_lint_pass!(AbsoluteSymbolPaths => [ABSOLUTE_SYMBOL_PATHS]);
+
+pub struct AbsoluteSymbolPaths {
+    pub absolute_symbol_paths_max_segments: u64,
+    pub absolute_symbol_paths_allowed_crates: FxHashSet<String>,
+}
+
+impl LateLintPass<'_> for AbsoluteSymbolPaths {
+    // We should only lint `QPath::Resolved`s, but since `Path` is only used in `Resolved` and `UsePath`
+    // we don't need to use a visitor or anything as we can just check if the `Node` for `hir_id` isn't
+    // a `Use`
+    #[expect(clippy::cast_possible_truncation)]
+    fn check_path(&mut self, cx: &LateContext<'_>, path: &Path<'_>, hir_id: HirId) {
+        let Self {
+            absolute_symbol_paths_max_segments,
+            absolute_symbol_paths_allowed_crates,
+        } = self;
+
+        if !path.span.from_expansion()
+            && let Some(node) = cx.tcx.hir().find(hir_id)
+            && !matches!(node, Node::Item(item) if matches!(item.kind, ItemKind::Use(_, _)))
+            && let [first, rest @ ..] = path.segments
+            // Handle `::std`
+            && let (segment, len) = if first.ident.name == kw::PathRoot {
+                // Indexing is fine as `PathRoot` must be followed by another segment. Similarly,
+                // `len() - 1` is fine here for the same reason
+                (&rest[0], path.segments.len() - 1)
+            } else {
+                (first, path.segments.len())
+            }
+            && len > *absolute_symbol_paths_max_segments as usize
+            && let Some(segment_snippet) = snippet_opt(cx, segment.ident.span)
+            && segment_snippet == segment.ident.as_str()
+        {
+            let is_abs_external =
+                matches!(segment.res, Res::Def(DefKind::Mod, DefId { index, .. }) if index == CRATE_DEF_INDEX);
+            let is_abs_crate = segment.ident.name == kw::Crate;
+
+            if is_abs_external && absolute_symbol_paths_allowed_crates.contains(segment.ident.name.as_str())
+                || is_abs_crate && absolute_symbol_paths_allowed_crates.contains("crate")
+            {
+                return;
+            }
+
+            if is_abs_external || is_abs_crate {
+                span_lint(
+                    cx,
+                    ABSOLUTE_SYMBOL_PATHS,
+                    path.span,
+                    "consider referring to this symbol by adding a `use` statement for consistent formatting",
```

Comment:
Some more `symbol`

We can borrow [the book's wording](https://doc.rust-lang.org/book/ch07-04-bringing-paths-into-scope-with-the-use-keyword.html) here since while I also find myself calling `use` a statement it's not actually one

```suggestion
                    "consider bringing this path into scope with the `use` keyword",
```

Reply:
We should probably clean up that up around clippy, there's quite a few usages of it in other lints